### PR TITLE
[RFC] Transition Frontier Bootstrap

### PR DESCRIPTION
**Pull request by nholland94**

### CODE REVIEW

File: rfcs/0010-transition-frontier-bootstrap.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+This data structure handles any number of root transition hashes `<= K`. As such, the interface of the datastructure should be such that there are three possible results from a query: `Known of hash`, `Unknown`, `Out_of_bounds`.
+
+### Root hash validation/punishment
+[detailed-design-root-hash-validation-punishment]: #detailed-design-root-hash-validation-punishment
+
+A `transition_frontier_root_hash` is valid iff one of the following conditions is true:
+1. `transition.transition_frontier_root_hash = genesis_state_hash && transition.protocol_state.blockchain_state.length < K`
+2. `transition.transition_frontier_root_hash = transition_frontier.root_hash_at_length(transition.protocol_state.blockchain_state.length - K)`
+
+In order to check conditino #2, the transition frontier now contains and maintains a transition frontier root history, updating it on root transitions. The transition frontier can be created with an initial history, allowing the program to pass the history accumulated during the bootstrapping phase to the transition frontier when it is initialized.
```
Comment (johnwu93) [inline rfcs/0010-transition-frontier-bootstrap.md]: conditino -> condition

**Comment by cmr** [comment]
> I'm not actually sure which function it should be. I will make it more generic in the RFC as this is an implementation detail which can be decided later.

log sounds about right, assuming each sample you make multiplicatively reduces the probability of getting it wrong. @johnwu93 and I are working out a model for this.

**Comment by bkase** [comment]
> This is a problem even with parallel sync ledgers, I believe. It's cheap for an adversary to trick you into tracking a new root, so they could just make you track a bunch of roots at once, and you are right back in the same situation. I still think we should think about this more because this is the big remaining issue with this RFC.

Could you put a note about this in unresolved questions?

**Comment by cmr** [comment]
log definitely isn't right, and the model we came up with answers a different question.

Regardless, the critical correctness property is that the probability of successful bootstrap to is no lower than the probability that a random peer is honest. Otherwise, attackers can inevitably fool all new participants in the protocol. From that perspective, it's sufficient to sample just one random peer and trust them.

**Comment by nholland94** [comment]
> Regardless, the critical correctness property is that the probability of successful bootstrap to is no lower than the probability that a random peer is honest. Otherwise, attackers can inevitably fool all new participants in the protocol. From that perspective, it's sufficient to sample just one random peer and trust them.

I think that holds for determining the initial root you want to target, but I'm not so sure if this holds when considering retargeting logic as any transition you read that has a greater length and a state hash you haven't seen yet will trigger a retarget, so even if the vast majority of the network is honest, a single adversary in your gossip network neighbors could trick you, though you would likely realize the trick soon.

File: rfcs/0010-transition-frontier-bootstrap.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+A bootstrapping node that wants to obtain the root corresponding to a proven protocol state `s` will request the root protocol state `root` along with an efficient ancestor proof: a value of type `State_body_hash.t list`. It verifies the proof by checking the list as a Merkle inclusion proof. This proof is about 200 kb (2000 field elements). A node only needs to request these proofs while it is in the bootstrapping phase. Once it has bootstrapped it no longer needs them. Furthermore, we could optimize it to save previously obtained inclusion proofs so that it would need far shorter inclusion proofs for subsequent updates.
+
+### Constants
+
+Thoughout the design, we will describe terms in relation to a constant `K`. The constant `K` is a natural number representing the transitions before finality can be assumed. As per Ouroboros proof of stake, `K = unforkable_slot_count = 2160`. The maximum length of a branch in the transition frontier is `K`, as the transition frontier root represents the point of finality of the network.
+
+### Root hash validation/punishment
```
Comment (imeckler) [inline rfcs/0010-transition-frontier-bootstrap.md]: this should be removed

**Comment by bkase** [comment]
@nholland94 should you update this or is it dead now?

**Comment by bkase** [comment]
@johnwu93 is commandeering it and rewriting to reflect the current design

**Comment by cmr** [comment]
@johnwu93 ping

File: rfcs/0010-transition-frontier-bootstrap.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+### Activation Conditions
+
+A node would want to bootstrap when it sees that is not synced to the network compared to its peers. It does this by comparing the length of its root's consensus state and the length of the consensus state of the transition that it currently sees from its peers, `seen_transition`. A node can see if the length of its peer's root consensus state is at least `seen_transition.length - K`. If this value is greater than  the node's current root length, then the node is out of sync with the network compared to its peer. This peer's root may exist in the node's frontier, so it would not be too out of synced and the node could run a catchup job to be in synced with the network (See [ledger_catchup.ml](src/lib/ledger_catchup/ledger_catchup.ml) for more details). It would very out of sync when this does not occur. As a result, a node would want to bootstrap when `seen_transition.length - root.length > 2k`.
```
Comment (cmr) [inline rfcs/0010-transition-frontier-bootstrap.md]: this `ledger_catchup` link should be `../src`

Comment (cmr) [inline rfcs/0010-transition-frontier-bootstrap.md]: When reading this I found it hard to keep straight node vs peer. It doesn't help that multiple meanings of "node" are in play. Maybe we can rewrite this with names for the protocol participants: Alice is trying to bootstrap, and she is getting the proofs she needs from Bob.

File: rfcs/0010-transition-frontier-bootstrap.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+An `external_transition` can be represented by its `protocol_state`, which contains a `blockchain_state` and `consensus_state`. An `external_transition` can be hashed by the hash of the hash of the `blockchain_state` and the `consensus_state` and its parent's hash. This hash is represented as the type `State_hash.t` and the hash of `blockchain_state` and the `consensus_state` is represented as the type `State_body_hash.t`. The hash of an `external_transition` is eqauted to :
```
Comment (cmr) [inline rfcs/0010-transition-frontier-bootstrap.md]: "equated", but I would rephrase, "is given by".

**Comment by nholland94** [comment]
Closing this as it is mostly outdated. We did end up implementing this design initially, but a lot has drifted in 2 1/2 years since this was written.
